File: sorting_algorithms/merge_sort/merge_sort.js
/**
 * Merge sort algorithm
 * Worst-case: O(n(log(n)))
 * Best-case: O(n(log(n)))
 * Average-case: O(n(log(n)))
 * Worst-case-space-complexity:
 *     - O(n) auxillary space
 *     - O(1) with linked list
 *     - Stack size is O(log(n))
 * Characteristics:
 *   - Stable
 *   - Not adaptive
 */

/**
 * Merges the right and left arrays
 * @param {Array} arr - original array
 * @param {Array} leftArray - left subarray
 * @param {Array} rightArray - right subarray
 */
function merge(arr, leftArray, rightArray) {
  let i = 0;
  let j = 0;
  let k = 0;
  while (i < leftArray.length && j < rightArray.length) {
    if (leftArray[i] <= rightArray[j]) {
      arr[k] = leftArray[i];
      i += 1;
      k += 1;
    } else {
      arr[k] = rightArray[j];
      j += 1;
      k += 1;
    }
  }

  // place remaining item on left array
  // into the original array
  while (i < leftArray.length) {
    arr[k] = leftArray[i];
    i += 1;
    k += 1;
  }

  // place remaining item on right array
  // into the original array
  while (j < rightArray.length) {
    arr[k] = rightArray[j];
    j += 1;
    k += 1;
  }
}

/**
 * Merge sort algorithm
 * @param {Array} arr - array of integers
 */
function mergeSort(arr) {
  if (arr.length <= 1) return;
  const mid = Math.floor(arr.length / 2);
  const leftArray = arr.slice(0, mid);
  const rightArray = arr.slice(mid);
  mergeSort(leftArray);
  mergeSort(rightArray);
  merge(arr, leftArray, rightArray);
}

function main() {
  const arr = [102, 101, 19, 48, 13, 99, 71, 13, 52, 96, 96, 73, 86, 7, 99, 100];
  console.time('mergeSort');
  mergeSort(arr);
  console.timeEnd('mergeSort');
  console.log('Ascending merge sort:\n', arr);
}

main();

File: sorting_algorithms/merge_sort/merge_sort.py
#!/usr/bin/env python3
"""
Merge sort algorithm
Worst-case: O(n(log(n)))
Best-case: O(n(log(n)))
Average-case: O(n(log(n)))
Worst-case-space-complexity:
    - O(n) auxillary space
    - O(1) with linked list
    - Stack size is O(log(n))
Characteristics:
    - Stable
    - Not adaptive
"""


def merge_sort(arr: list[int], size: int):
    """
    Merge sort
    """
    if size <= 1:
        return

    mid = size // 2
    right_arr = arr[0:mid]
    left_arr = arr[mid:]
    merge_sort(right_arr, mid)
    merge_sort(left_arr, size - mid)
    merge(arr, left_arr, right_arr)


def merge(arr: list[int], left_arr: list[int], right_arr: list[int]):
    """
    Merges right and left array by placing the items
    in the right slot
    """
    idx = 0
    left_idx = 0
    right_idx = 0
    left_arr_size = len(left_arr)
    right_arr_size = len(right_arr)

    while (left_idx < left_arr_size and right_idx < left_arr_size):
        if left_arr[left_idx] < right_arr[right_idx]:
            arr[idx] = left_arr[left_idx]
            idx += 1
            left_idx += 1
            continue
        arr[idx] = right_arr[right_idx]
        idx += 1
        right_idx += 1
    while (left_idx < left_arr_size):
        arr[idx] = left_arr[left_idx]
        left_idx += 1
        idx += 1
    while (right_idx < right_arr_size):
        arr[idx] = right_arr[right_idx]
        right_idx += 1
        idx += 1


if __name__ == '__main__':
    arr = [102, 101, 19, 48, 13, 99, 71, 13, 52, 96, 96, 73, 86, 7, 99, 100]
    arr = [i for i in range(25, 0, -1)]
    print(arr)
    merge_sort(arr, len(arr))
    print('Ascending merge sort:\n', arr)
